extensions/control-plane/transfer/transfer-data-plane-signaling/src/main/java/org/eclipse/edc/connector/controlplane/transfer/dataplane/flow/DataPlaneSignalingFlowController.java

```diff
@@ -97,46 +99,66 @@ public boolean canHandle(TransferProcess transferProcess) {
+            return clientFactory.createClient(dataPlaneInstance)
+                    .start(dataFlowRequest)
+                    .map(it -> DataFlowResponse.Builder.newInstance()
+                            .dataAddress(it.getDataAddress())
+                            .dataPlaneId(dataPlaneInstance.getId())
+                            .build()
+                    );
+        } else {
+            // TODO: this branch works for embedded data plane but it is a potential false positive when the dataplane is not found, needs to be refactored
+            return clientFactory.createClient(null)
+                    .start(dataFlowRequest)
+                    .map(it -> DataFlowResponse.Builder.newInstance()
+                            .dataAddress(it.getDataAddress())
+                            .dataPlaneId(null)
+                            .build()
+                    );
+        }
     }
 
     @Override
     public StatusResult<Void> suspend(TransferProcess transferProcess) {
-        return selectorClient.getAll().stream()
-                .filter(dataPlaneInstanceFilter(transferProcess))
-                .map(clientFactory::createClient)
-                .map(client -> client.suspend(transferProcess.getId()))
-                .reduce(StatusResult::merge)
-                .orElse(StatusResult.failure(ResponseStatus.FATAL_ERROR, "Failed to select the data plane for suspending the transfer process %s".formatted(transferProcess.getId())));
+        return onDataplaneInstancesDo("suspending", transferProcess, DataPlaneClient::suspend);
     }
 
     @Override
     public StatusResult<Void> terminate(TransferProcess transferProcess) {
-        return selectorClient.getAll().stream()
-                .filter(dataPlaneInstanceFilter(transferProcess))
-                .map(clientFactory::createClient)
-                .map(client -> client.terminate(transferProcess.getId()))
-                .reduce(StatusResult::merge)
-                .orElse(StatusResult.failure(ResponseStatus.FATAL_ERROR, "Failed to select the data plane for terminating the transfer process %s".formatted(transferProcess.getId())));
+        return onDataplaneInstancesDo("terminating", transferProcess, DataPlaneClient::terminate);
     }
 
     @Override
     public Set<String> transferTypesFor(Asset asset) {
-        return selectorClient.getAll().stream()
+        var result = selectorClient.getAll();
+        if (result.failed()) {
+            return emptySet();
+        }
+
+        return result.getContent().stream()
                 .filter(it -> it.getAllowedSourceTypes().contains(asset.getDataAddress().getType()))
                 .map(DataPlaneInstance::getAllowedTransferTypes)
                 .flatMap(Collection::stream)
                 .collect(toSet());
     }
 
+    private StatusResult<Void> onDataplaneInstancesDo(String name, TransferProcess transferProcess, BiFunction<DataPlaneClient, String, StatusResult<Void>> action) {
```

Comment:
micro-nit: would rename `name -> action`

Reply:
ok, I'll need to find a different name for `action` then :joy:

Reply:
ah shit, overlooked that :D